libsolidity/formal/CHC.cpp

```diff
@@ -1101,186 +1085,173 @@ pair<CheckResult, CHCSolverInterface::CexGraph> CHC::query(smtutil::Expression c
 	return {result, cex};
 }
 
-void CHC::addVerificationTarget(
-	ASTNode const* _scope,
+void CHC::errorConditionEncountered(
+	ASTNode const * const _errorNode,
 	VerificationTarget::Type _type,
-	smtutil::Expression _from,
-	smtutil::Expression _constraints,
-	smtutil::Expression _errorId
+	smtutil::Expression const& _errorCondition
 )
 {
 	solAssert(m_currentContract || m_currentFunction, "");
-	SourceUnit const* source = nullptr;
-	if (m_currentContract)
-		source = sourceUnitContaining(*m_currentContract);
-	else
-		source = sourceUnitContaining(*m_currentFunction);
+	SourceUnit const* source = m_currentContract ? sourceUnitContaining(*m_currentContract) : sourceUnitContaining(*m_currentFunction);
 	solAssert(source, "");
 	if (!source->annotation().experimentalFeatures.count(ExperimentalFeature::SMTChecker))
 		return;
 
-	m_verificationTargets[_scope].push_back(CHCVerificationTarget{{_type, _from, _constraints}, _errorId});
-}
-
-void CHC::addVerificationTarget(ASTNode const* _scope, VerificationTarget::Type _type, smtutil::Expression _errorId)
-{
-	solAssert(m_currentContract, "");
-
-	if (!m_currentFunction || m_currentFunction->isConstructor())
-		addVerificationTarget(_scope, _type, summary(*m_currentContract), smtutil::Expression(true), _errorId);
+	bool scopeIsFunction = m_currentFunction && !m_currentFunction->isConstructor();
+	auto errorId = newErrorId();
+	solAssert(m_verificationPreTargets.count(errorId) == 0, "Error ID is not unique!");
+	m_verificationPreTargets.emplace(errorId, CHCVerificationTarget{{_type, _errorCondition, smtutil::Expression(true)}, errorId, _errorNode});
+	if (scopeIsFunction)
+		m_functionVerificationPreTargetIds[m_currentFunction].push_back(errorId);
 	else
-	{
-		auto iface = smt::interfacePre(*m_interfaces.at(m_currentContract), *m_currentContract, m_context);
-		auto sum = summary(*m_currentFunction);
-		addVerificationTarget(_scope, _type, iface, sum, _errorId);
-	}
-}
-
-void CHC::addVerificationTarget(frontend::Expression const& _scope, VerificationTarget::Type _type, smtutil::Expression const& _target)
-{
+		m_functionVerificationPreTargetIds[m_currentContract].push_back(errorId);
 	auto previousError = errorFlag().currentValue();
 	errorFlag().increaseIndex();
-	addVerificationTarget(&_scope, _type, errorFlag().currentValue());
 
-	m_context.addAssertion(
-		errorFlag().currentValue() == previousError ||
-		(_target && errorFlag().currentValue() == newErrorId(_scope))
+	// create an error edge to the summary
+	connectBlocks(
+		m_currentBlock,
+		scopeIsFunction ? summary(*m_currentFunction) : summary(*m_currentContract),
+		currentPathConditions() && _errorCondition && errorFlag().currentValue() == errorId
```

Comment:
I'm actually not sure we need `currentPathConditions()`

Reply:
Me neither, but it was there before.

Reply:
I think it can be removed.